### ISSUE TITLE
[no-unused-vars] Throws warning that generic type parameter is unused if an inferred type uses the same name

### ISSUE THREAD

**Issue by AllySummers** [typescript-eslint/typescript-eslint#3578]
- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**
```TS
export type GuardedType<T> = T extends (x: unknown) => x is infer T ? T : never;
```

**Expected Result**
No warnings should be given as the type parameter is used.

**Actual Result**
ESLint gives the following warning
```
(type parameter) T in type GuardedType<T>
'T' is defined but never used. (@typescript-eslint/no-unused-vars)eslint
```
**Additional Info**
If you remove the `<T>` from the type, TypeScript will then throw the error "Cannot find name 'T'".

Apologies if I have missed anything or made a mistake in filing this issue.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.28.0` |
| `@typescript-eslint/parser`        | `4.28.0` |
| `TypeScript`                       | `4.3.4` |
| `ESLint`                           | `7.29.0` |
| `node`                             | `14.16.0` |

**Comment by bradzacher** [comment]
The bug here is that the reference resolver runs after the variable declarer.
So we analyse and find `infer T` and create `T` in the conditional type scope, and then after that we resolve the `T extends` reference - which is within the conditional type scope, so it references that.

This is actually pretty complicated to handle.

---

Separately from the bug.

You really shouldn't be writing code like this as it's really hard for people to understand.
When you use `infer T`, TS actually declares a new type `T` in the scope of the conditional expression.
What that means is that you've shadowed the original type `T` within the conditional scope.

In a nutshell it's the same as doing this:
```TS
function foo(x) {
  function bar(x) {
    return x;
  }
  return x;
}
```

It's hard for someone to track which `T` refers to what declaration of `T`.

If you rename the inferred type to a different name, you'll have clearer code and avoid the bug altogether.

**Comment by AllySummers** [comment]
Thanks for the response @bradzacher. I previously modified my code to use a different type name for the inferred type, I just thought it would be considered a bug so I reported it.

If it's too complicated or you just don't want to fix it, feel free to close. :)

**Comment by RA80533** [comment]
@AllySummers Could you make a quick Playground link with your type guard? I'm curious to see how it's used.

**Comment by AllySummers** [comment]
@RA80533 
[Here you go](https://www.typescriptlang.org/play?target=99&jsx=0&module=1#code/PQKgBAUgrgzgLmAZgewE5hgCzXAdgUxhjBGACgBjZXeMASxgEFVUBDATzAF4xm32AdAz4cA3GTJ1ccfKkSsK+XgBNlqQsQDeZMBjjr8cAFx7UUgOYAaHWABedAA4n4Z3FZsU6cds-0XrulRQ0qg+phZkAL4SUjJyCkoACtgEAHJQALYARrJg2rreDvgmAOTYGfglYAA+YCUA7mgA1lW1JRnIWXQANpUBYLiZOai+ruZRMdKy8opgAMp0Wd0WeTa4rBWj-jas5sUDQ7ITklPxs4myMNSruoh0qPCpG-su27rdrI-PW279u-uDbKyfodOCYWQAfh+7lurDBkOhf1U6iIJkYyI0-QcKUIaJYHAAPMlqPh0kDUAA+fowRbLNwwEwLJYrWoidgEpl08wU46IYIUOB0a7CDFEAAUXnwGRMwSauGQ9VwAEoTJKMvRiOi1BobmB1HAoKhcBqAPJZABW+AFEpkGSVNl0ADJHRq5n43DapRCBC58IZ7bonS6GG6xp6Mt77A4A4Hna73eZw97PN4Y0H42G1cnkMF9Ox7dEyHzcAKhcaQ7SLOGZbg5QrlarbRr5pW3Lr9YbyzAzZbrWq02A4yGE0mBOsKgOhzAycNR-8CxIqDQEAxiQQACrsIrcMDLOKsbqMZafMVlZAVEqWOqNVAtK-tTo9SpK8RL2gV5luACyHBybJ3GS-vgbJHnQJ4flyL4SMWpbCjAa6kocqDVmAsryoqKr0E2DBgAhM65PkeqGJ2poWlacDhpOLoSvBOKbkUo6FPgSpURmVZZmOSELpQ1DvrRJL4ag-48Kw+LsKB4H8WkXHiEW-KCnBFwPNQKFofWmFqs2SlXMahEdkapG9hR-YOoOwYwKG7G2t6dwPHATwTqZU6WR6HEfF8jmBmZGqCXOeyscIuDsCaiCjqC4KoFeDAAKq4Mo+B3AQyhRRZCYBUwQUhaO8jwpFGqxfFiX4MlbFuOlWoojAc6ijA6V4UhbKjtiJK1U55mchYP7sH+YmjjSn7mK1hagOAAAS+DdEU6DmFAonKMQpBySWCldj25GqbW6ENlhUrNgASlaaDKESqDIFN3gANL4OwV6sEFFLtsRBlMcgiA7eqXCfXUnRGVUbVgAAhMIvUmV5Q5NgDX2DN03T-Rxb76FAApoNwX1rQKvLyWWGoAELIMgvR3RtdYYY2u04Vk+OE7pNj6caL1vZpn08CUlME-gd0lJjy3Y8Oma2jWJPbZpOGvG2elPfTW74K972oyzYvmFzhYwStPlIcTW0adhxCAsMj0Gs90uy0zX0lHrsjK9BWNwQAYjbuCa+pZPqjh9s89cEuG1LRQm02zN1KrZZW0tsFdukMNO6Tcs4dD3QGyRDNy5DLPBIVUjFVU4O7QHcfc2H+VxQlGfKFHws66hRdFcoCdG77jP+2bafF0lIcjWAOMTQqzbp7gXhlgt5BkPgAAeDg4GATFgCaQliQS64PTw641Lwc8L+Io-j6gCBTwA4rNqDxco9H4PPi9gMvo8yHFxBiiPJh3fm3APSPzYSrgiC5NFSpgBCYDRWAEwBAABusgN5jwnnvA+x0F47hQo-H+XAHoi2IOucBW8d7SzAAAGUlGweOPBFYrwtugWobNqbEKgDDFezdq6yRGjocAAABeUMheDGlEvwMA9QvCYDsLIZAYAJpSnwNIBaIAeEwxAOABwnxiDwknoQOAjCwBMPijACgZgHBq3XEo+R4IwDmDoKA40AADThpjJ6CJyGAWWU87o1zpvI1AUAlAo3kN0GA+ABAqPXJgHCQc4JgFAaEDAdAMg9FElYvITC6RNA1GyEKiRIhCJHoobRk9MBwiwrY3A3R2AqIUOkuAxBWBgAoAeboWQFDxJkLQBxWESjEDppPLBOEOjqDAEUw0cJvEqKYbItg6pOGry4XAQRdSEA8LBJkpQRiTGtKKP0wZGxFG0D8UoSZ0TWAOAcPk7Z1DhEVDEfQemBjOEcEYeQTeE9AldkSYgeeV4AAaqTr7zTAPAoKiDkEVwXmKThD8gpXkmSYZ5mERk4TZASfec1ionwJM8ikD0vYkWBvwAFLAf5xk4QIfAIT2BikmVBYa0j+msKUAASTego+ZoiY6lI4WJbhvD+GnSEb0Y5JSooIGkZI7o0iwCyKILMtZyiSCqPUZoxwOi9GirpWY4BB5LHjLADYuxWCUZ3S6cyjV24Gn8qIt7SeLi3HoA8V4nxEqBmiVWUq+OAA1A8rjolbOmXw2lxj6WmMmaY5Ztr1RbN0bQVVOy9mcFVfgXhuQFGmPtZYiUb1WFdKZfwJUV4UYcwoHwo5oiVy4BUaKi5nA6BvTjcq5s2ri1XOHhA7eSAHYai6j1fgIUnlgFeVfURHyvlPyQQyi+FIxT2qBTdMVYLML2ubDPaFsLD7wuloi5Ftd6ZKOHQeAc1Rajoo4Fle1IKlHcQYdailF8DHdGoOYXcDAECy2QCEi9rAPmxytBoUSnAUDoCoCwciiyLBWuYVKrRsr4DEFLfKr1Zi1SWMAnAbNhAU2cDsQY0xM05owEsQ4U6wC6BH39UMuWGy5aqsmfh1ZaHD6alTRwWxb0KMfJI0o0VpjoNdPMKwKQ8ABBgAAMLapsWUwCr9ZYAEYAAMaqEpoE2VuFYOEL3wBrTc+tdyEmZUeYwRgQ61SjtuqwEwc7jqaYpNrcmxBDMLqKASJOrBWAUnQbcxtgVgoacYFeRgONtMC0Q3pgz0CCTGdulkPzc0AuedM67cz0DLOnxs3ZleFnj6Lps1keztaMENo9vc9TAW3O8Bxu5njXmpS6a6fpsAhnctUi6cFir-mPPVYUCFw+AWisRebIlhFcWHq1E68l42b1WCpYS9FpLVmbMUDS8phAqnnNts0+5grvAePuYACLFelD5srzWjNaaCztsLjWKAHcYEV26ygTvrfazhPr42BtlZ63VuFY3Yv3aG4927r365dMmyN57XW3vKCm3WmbTmMoudy4twra33MAFENuldsydvbNWTuedusdp7LXTuNYu1joz63br7Eq4weH12ov-f6992zH3RsA+p8N3rdOqcy0G79pnlO7vU6B39+dL3rNvfwMDjLs3wfzbyx56HvBVtw-c7bBHW2kf46q-t5XDWMcnbO10vHJPCddOJ-V+Ht1EAnfl+T-HMWBfU-ixzvn9PWc1dp5zr7DuFBO7tyz2WT73dH3t17oXvPfee8G4gYXjmstqYhwt-LUvGAy94LDuX7nd4K8fr5tXKOhto6O5r3Hl3GsG9C6TxrJu1fy9uuYE7Kfzefat67m3Fv+cpZ95bibLem+A-b37wbAfbdB65670PgfW9ve5A5lTYPGA5ej5L5bMOE9J94Lvdzo1U-Au2xnxrtWSfo5+7n87+eicnaN100vJPy9sar41zAJ3V81+ZwPr3Dfa-N+Hx36n7PG-d+1134P+vf+P4h4AEu5e7chv7f6sCYBh4T4R5zauZQ5z7S6y68C2zJ4r7uaUpr5jpK4k6Z7b71a75NZq5a5PqH767H4l6m6NaV5q4p63Q35q6r63R0AnaYH37O515P7AGcGDaM5f5-5u7gECE8594j7U6978GAGn7cE2ZgGiHv6u5QFCFSEcbQGg6wFi7wEx6IFx7IGMCoFL7oG8CUruYQBYHp64Fb7Z4a7EF55q566sCF7Y4n6sBn71YX67BX70G36NYsFq6YG3TmgnZmHsEe4qHP4P4gG8EyFvaf4v6d7KFRH-6JE8HSEpGyExHU5KHyEQF0CZGu7mhqGZYFxwG9pXgCAVH0YMijKEiVaPzIqhH96nz1Hj7qElGaFlFgAVECBVF4j8Awr+b1EmYuwdaREEj1ErqGLQIwA+jnj4Bij0bPxTFzSUTcTTbFFqwMC4L7jdC5ZvLdrEDbGyAHgK5qSKjp6abm6aatEbG8wwBHH4J7Fdo3w4J4IHiLb7EvEPEnFlwXF5ZZ75ZXGMArweY3Gi7fG7GaafEfIQkfHPEwlvHdCFbQmHGImnGbT1h-Gq6z5EGnZAkgk4wgk8Zglg4QlPEjzvKok7FwkUkHGvHUnLYon0nHFInS5MkQnolCxYmo7aG4krba5oirb4msiEmsg8YgmrYkkaFklQnwlUksk0mUnMn4LIlynKnvFslqmwkJ7slom-Eb7R4Ak4mY6nYH5IHkEJ7Cn5ZEkSkgmw5SntEynAlamImKl0namna6kMlx5ekKk6kunem2y+n4KclbTclGlLZ8lmm6EWmk7G5oi2xWkeY2msirZ2kgm2wOmbH3GInklKkemEkBl+melFkqmam0lfGun+kVkImBnBkamMC7z1ndChmYkGn-G1bGloj8mkHmmOFoiJ6n4JkV5oi7xJmimMmpnpmshBmsi7xZl3FOnNlumVnenimlkNlpkbmsmk7LkoF7mNkHmjTNmtnnHtnYmRkmk9l44xn9kL5DkoEjmGFdIMGMCjTjkplsmsiw4ZkglNmsijQLlwRLnbkrm1nFnrk1nyllk+mgXVn5lVn6EHlNlwVvkHmUonn6lK6GmdmXndnRnx53lxkPn6FPmNneG8CjTMFoiUoflim2nfm-lzkgnHmsiUpAVdggVQXqk7mFncUemQUIXelbn8WIU-moVBmoUoWiXenHmoUYWoUQCYU6aVxcnnk8ldmIG9m3n7DEWuHDmX7PmQFohUVdJ+GMAmFdJBG8AQB0WTlfk6kzl-ksUgkYWsgQAcUahkn-IqVnG4DlEVF7gsnVHQoLzm5oLpbh6Om5k+XeZ+UBUCBBX4IhVrzDEDrLyooGRJUHgzFXAVBijZUEK-I5xfSFVrEg63HXDFoSQwBnxEpKImC9o-LpU-yZXGhNUjFQqpVLE7oubhgHrwAkoSDrGqaATdTARiQ1V1WgqfIqUII9V-KtW0ySyzXeYIKdXEAzrdX9oMDNoTWtqhRqgDVwBHpkonrIBsJEYXptjLC0B3oPrIBPpga6yvpEDvpIAozfrqACh-puAAaSqEDSraLYw8bqC9KlIDD4D1DlLnhdBJTLGHwsozJfW-qFArCfqGKQZgCobTF+okDXIVWqbBBlhTXGYArlaWGYRihNULVmaSFJE05QQjWNrE3UCk0S6ebk3I5WFq7haNU007UVzxHW4rzN5M2E0s19xs3Hi1Uz5LY45c2b6q474562FU0C3FWRb02pG2ai1vZZB60f4maeWs24Ds0IGmnS5Do4H1Z4HWF762Fmm65pXU1zXfK01a3C316G2u4G21ATY+1e5A7i0i6S0k0y2Q7aGW0xnF6K2WHK0EGq0k4kE671YOFOFGZk781u19qa2jEcFxaB28FF0-Yl1Ppl1C4h1RVqym3m1R38kx2Dn6HW0U22080q02HJ12HO1H5q4uFuFF5m7Z1rXu2C101e1P5l1+2LKu4UBl01z+2C5l2h5V0wEFy10R1y2x7x7EWkVL4t3c0J1F6EGY5d1O1p0F4UHxll7UFX7q053NUoLa2F2L0M5l1z2v2u4L0z3+7L1l3cir1tE11S1m2b0c3b16EGHkWUUH1K08kd0O1n0Cn2GX192UE31Pkk50Evk+H30j250DoT2DZlKf1e7T0B2kODbf02ZKCUOn7-1l1QGAOVXGgb1gSy3gM6E71N1QPL6UUYGwPx3wOJ2d31Yp1kF3kk791UGYP1bYPGWMG+GsEu0a2ENjEv0-3F10MKDz0V1-3aNXraN8LaN5HMNE0gN12z7R3cOL7QNvkYGmGCNt1H3Y4n377IM90WlSPoPn6320HX4+HUX+GNbWUk4hHD0laIaP1C3qNvYkOaM1bv26PaO0MJOuEMNGNl10Bl2FFmNh3S3sNijdG9E1Hsh1H3QqMP0e351hHNH3R5MR5sMnjFPTF9G1GDEVN4ORPzVj2e1jH1FLW6AtIdVqMcETE7UdFHVdGVHTHlWh0R6FXs2YXYWMBU2+UYkYTVNQqMCeWLNgMHl8VCUsmwM4Vog4xrNxUbPKhbOajAkim7OIlLOoWHPumIWCWvM7EnMdlnM2F4mNXrNCzRN01QkilEkPM7FPMyXFkvOrkQUHkiVHMhmK2nO8lXkEUXORN+VAta0gvWn0WpngssmQuIsNkwvgUwXvOwswUIsfPFniVQtIsrMXm-PXmCm92k4YubZYs3O8B3N4v2U+nfmEv4LEu0swVkvQUNmUvkubkHn0sks7mSUMs-FMsaV4VaU3mEW6WDn6UoGcuCxaw8u4vJn4sOW7kznCsHiitUuksHnSuSs7k0s2s7nytisNlKsKvJ7LPlYouaWW3aVasDnX172eFL76uqWGu9PNjGsTmelTmMVOVzmWuHj7PPN2vwtysHketus7nSWeuUXetojfOov4XIM6VBskUGGhvQMKNvnhvctRvbMEmfmCuOX7nMUAXJvWsyu8XpuoVOs9tw5ZvIVHnoWFu8vMsO3+uavss6ul4hs0E1uvmmUcY0X1tXPYvRt8smsCtpkJvttL4uVsVdupvKu9uoX2s8VraZsSUjuoVyVnsYEHlKXblfOTtRlluBv3m6sLujkUX2NmU0WBFoi2X-OXOAtGvbuxvinxtttIUdsFtsUgkeWRVr1qx7OFPdGFUpX9EcnrsQeNsOsnuYeBWIk4eEh4dgeYsbs8sQmTFlzm50cTMQn9XTOJVkcFhAA)

I use the `GuardedType` to create "aliased" or "helper" type guards, so that I only ever need to define a type guard to check an object or whatever and I can use it like `isArrayOf(item, isSomething)`, or I can create an alias for checking if it is a combination types, or an array, or either the base item or an array of that item and so on. 

In one of my apps I had places where I could accept an array of something or maybe an array of something, and creating a type guard to for each of the base item, an array of the base item, or maybe an array of the base item for some 20-30+ types would have been repetitive and useless, instead of using `isArrayOf` or `arrayAlias` or something to that effect.

Please let me know if you find it useful or have any suggestions :)

**Comment by Josh-Cena** [comment]
This does not seem to reproduce in the latest version ([playground](https://typescript-eslint.io/play#ts=5.2.2&fileType=.tsx&code=KYDwDg9gTgLgBDAnmYcDiBXAhlAJsXAFWWAB5CA%2BOAXjkLlBmADtcBnOAChAC44NmAa2YQA7swCUNKiDgBLDnOYAzYFDgBVOAH5NcPs2AA3NQG4AUEA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tidmUQAmtAG4BDaKgyRE0aB2iRwYAL4hlQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)).
